File: LeetCode/JavaScript/0094 - Binary Tree Inorder Traversal.js
/*
94. Binary Tree Inorder Traversal
https://leetcode.com/problems/binary-tree-inorder-traversal/
*/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */


 var inorderTraversal = function(root) {
    
    const node_values = [];
    
    function dfs (node){
        if (node==null || node==undefined)
            return;
        
        dfs(node.left);
        node_values.push(node.val);
        dfs(node.right);
    };
    
    dfs(root);
    
    return node_values;
    
};

File: LeetCode/Python/0094 - Binary Tree Inorder Traversal.py
'''
94. Binary Tree Inorder Traversal
https://leetcode.com/problems/binary-tree-inorder-traversal/
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:

        if root is None:
            return None

        nodes = []

        def inOrder(root):

            nonlocal nodes 

            if root is None:
                return

            inOrder(root.left)

            nodes.append(root.val)

            inOrder(root.right)

        inOrder(root)

        return nodes
